bcex/examples/simple_strategy.py

```diff
@@ -91,25 +131,157 @@ def live_candles(self):
         return self.get_latest_candles()
 
     def get_candle_df(self):
+        """Gets df of all candle data
+
+        Returns
+        -------
+        df : pd.DataFrame
+            dataframe of candles
+        """
         live_candles = self.live_candles
         if live_candles.empty:
-            return self.historical_candles
+            return self.historical_candles.copy()
         min_time = live_candles.iloc[0].name
         historical_candles = self.historical_candles[
             self.historical_candles.index < min_time
         ]
         df = pd.concat([historical_candles, live_candles]).sort_index()
         return df
 
-    def place_order_at_crossover(self, df):
+    def is_new_candle(self, candles):
+        """Checks if there is a new candle in the dataframe
+
+        Parameters
+        ----------
+        candles : pd.DataFrame
+            dataframe of candles
+
+        Returns
+        -------
+        is_new_candle : bool
+        """
+        last_timestamp = candles.iloc[-1].name
+        if last_timestamp > self.latest_timestamp:
+            return True
+        return False
+
+    def make_candles_heikin_ashi(self, candles_df):
+        """Converts candles to heikin_ashi candles
+
+        Parameters
+        ----------
+        candles_df : pd.DataFrame
+            dataframe of candles
+
+        Returns
+        -------
+        candles_df : pd.DataFrame
+            dataframe of heikin_ashi candles
+        """
+        candles_df["ha_close"] = (
+            candles_df["open"]
+            + candles_df["close"]
+            + candles_df["high"]
+            + candles_df["low"]
+        ) / 4.0
+        candles_df["ha_open"] = (
+            (candles_df["open"] + candles_df["close"]) / 2.0
+        ).shift(1)
+        candles_df.loc[candles_df.ha_open.isna(), "ha_open"] = (
+            candles_df["open"] + candles_df["close"]
```

Comment:
is it the open as in the heiken ashi deefinition of open or the actual open ?
You replace the original candles so here it will be the actual open the first time but then will be replaced by the Heiken aki opens for the rest

Reply:
as I see you're doing a copy at the top,that's fine then